myia/compile/nnvm.py

```diff
@@ -0,0 +1,293 @@
+"""Linear implementation using NNVM."""
+
+import numpy as np
+
+import nnvm.compiler
+import nnvm.symbol as sym
+import tvm
+from nnvm.compiler import graph_attr
+from tvm.contrib import graph_runtime
+
+from .utils import get_outputs
+
+from ..dtype import type_to_np_dtype, ismyiatype, Array
+from ..prim import Primitive
+from ..prim import ops as P
+from ..ir import Graph
+
+
+SIMPLE_MAP = {
+    P.scalar_add: sym.elemwise_add,
+    P.scalar_sub: sym.elemwise_sub,
+    P.scalar_mul: sym.elemwise_mul,
+    P.scalar_div: sym.elemwise_div,
+    P.scalar_mod: sym.elemwise_mod,
+    P.scalar_pow: sym.elemwise_pow,
+    P.scalar_floor: sym.floor,
+    P.scalar_uadd: lambda x: x,
+    P.scalar_usub: sym.negative,
+    P.scalar_exp: sym.exp,
+    P.scalar_log: sym.log,
+    # This is not right tangent vs hyperbolic tangent
+    # P.scalar_tan: sym.tanh,
+
+    P.scalar_eq: sym.broadcast_equal,
+    P.scalar_lt: sym.broadcast_less,
+    P.scalar_gt: sym.broadcast_greater,
+    P.scalar_ne: sym.broadcast_not_equal,
+    P.scalar_le: sym.broadcast_less_equal,
+    P.scalar_ge: sym.broadcast_greater_equal,
+    # P.bool_and: sym.logical_and,
+    # P.bool_or: sym.logical_or
+    P.bool_eq: sym.broadcast_equal,
+
+    P.scalar_to_array: lambda x: x
+}
+
+
+def nnvm_bool_not(c, arg):
+    """Implementation of boolean not."""
+    zero = c.make_constant(0, nnvm_type=nnvm_type_map(arg.type))
+    return sym.broadcast_equal(zero, c.ref(arg))
+
+
+def nnvm_distribute(c, v, shp):
+    """Implementation of distribute."""
+    nv = c.ref(v)
+    assert shp.is_constant()
+    shp = shp.value
+    vshp = v.shape
+    if len(shp) != len(vshp):
+        # We need to pad the shape
+        vshp = (1,) * (len(shp) - len(vshp)) + vshp
+        nv = sym.reshape(nv, shape=vshp)
+    if shp == vshp:
+        return nv
+    return sym.broadcast_to(nv, shape=shp)
+
+
+def nnvm_dot(c, a, b):
+    """Implemetation of dot."""
+    na = c.ref(a)
+    nb = c.ref(b)
+    return sym.dense(na, sym.transpose(nb), units=b.shape[1], use_bias=False)
+
+
+def nnvm_array_map(c, fn, *array):
+    """Implementation of array_map."""
+    assert fn.is_constant()
+    fn = fn.value
+    if fn in SIMPLE_MAP:
```

Comment:
`c` doesn't have a simple_map attribute.  I'm not sure I want to provide one.

Reply:
I would add it for consistency's sake.

Reply:
The current code is basically a giant hack because I don't properly support graphs.  It will go away when I do it.  So I don't mind a little bit on unclean code there.

Also, there will never be a different map here.

Reply:
Maybe it shouldn't be a parameter to `NNVMConverter`, in that case.